plugin/trino-kafka/src/main/java/io/trino/plugin/kafka/encoder/protobuf/ProtobufSchemaParser.java

```diff
@@ -95,21 +98,24 @@ private Type getType(FieldDescriptor fieldDescriptor)
         return baseType;
     }
 
-    private Type getTypeForMessage(FieldDescriptor fieldDescriptor)
+    private Type getTypeForMessage(FieldDescriptor fieldDescriptor, Set<String> processedMessages)
     {
         Descriptor descriptor = fieldDescriptor.getMessageType();
-        if (fieldDescriptor.getMessageType().getFullName().equals(TIMESTAMP_TYPE_NAME)) {
+        if (descriptor.getFullName().equals(TIMESTAMP_TYPE_NAME)) {
             return createTimestampType(6);
         }
+        if (!processedMessages.add(descriptor.getFullName())) {
+            return UnknownType.UNKNOWN;
```

Comment:
how about throwing `TrinoException` about the cycle in the types?

Reply:
ok. pls see my comments: https://github.com/trinodb/trino/pull/15312#discussion_r1040836832